### ISSUE TITLE
Duplicate fields are generated into Fragment interface

### ISSUE THREAD

**Issue by gergely-ujvari** [apollographql/apollo-codegen#73]
I'm using version `0.10.9` and ran into the problem that some properties are generated more than one time into a Fragment interface.

I've assembled a very trivial example for this.
Schema: 
*zoo.graphql*
```
interface Feedable {
    _id: ID!,
    kind: String!
    feedWith: String!
}


interface Reptile {
    scaleSkin: String!
}

type Crocodile implements Feedable, Reptile {
    _id: ID!,
    kind: String!
    feedWith: String!
    scaleSkin: String!

    victims: Int!
}

type Iguana implements Feedable, Reptile {
    _id: ID!,
    kind: String!
    feedWith: String!
    scaleSkin: String!

    age: Int!
}
```

*root.graphql*
```
type Query {
    getZoo: [Feedable!]!
}

schema {
    query: Query
}
```

And now I setup a very simple query with some fragments:
*zooQuery.graphql*
```
query getZoo {
    getZoo {
        ...FeedableFragment
    }
}
```

*FeedableFragment*
```
fragment FeedableFragment on Feedable {
    _id
    kind,
    feedWith

    ... on Reptile {
        scaleSkin
    }

    ... on Crocodile {
        victims
    }

    ... on Iguana {
        age
    }
}
```

And now I check the generated interfaces and I have this:
```
export interface FeedableFragment {
  _id: string;
  kind: string;
  feedWith: string;
  _id: string;
  kind: string;
  feedWith: string;
  scaleSkin: string;
  victims: number;
  _id: string;
  kind: string;
  feedWith: string;
  scaleSkin: string;
  age: number;
}
```

As you can see, the `_id`, `kind`, `feedWith` fields are repeated three times, instead of one.
What's more, I think that the `scaleSkin` and `age` properties should be optional properties.

**Comment by gergely-ujvari** [comment]
I have edited the description, because of a copy-paste typo some parts were copied twice into the description.

**Comment by martijnwalraven** [comment]
I don't know too much about the TypeScript target, but maybe @rricard has a better idea what might be going on here.

**Comment by clayne11** [comment]
I don't think it should be optional. It should be a [discriminated union](https://www.typescriptlang.org/docs/handbook/advanced-types.html) using `__typename` as the discriminator.

**Comment by lewisf** [comment]
It looks like the intermediate representation consumes the top level fragment fields into the inline fragments, and the flow/typescript targets currently do a concatenation of the fields without field deduping on inline fragments which is why the flow/typescript targets.

a discriminated union in typescript (referenced as a disjoint union in flow) looks possible with the current IR. i might be able to give this a shot but am going to have to familiarize myself with the code a bit.

**Comment by lewisf** [comment]
it looks like the only way to use discriminated unions appropriately is if the __typeName field is actually requested by clients in the actual query since the field actually needs to be there for client code to `switch` on. This seems like a stricter requirement than what's required by swift? cc @martijnwalraven

**Comment by martijnwalraven** [comment]
@lewisf: The compiler [adds `__typename` automatically, except on root fields](https://github.com/apollographql/apollo-codegen/blob/master/src/compilation.js#L401-L421). We currently even [disallow clients from requesting it explicitly](https://github.com/apollographql/apollo-codegen/blob/master/src/validation.js#L35-L47). Originally we only added it when the parent type isn't an object type (because it is strictly taken superfluous if it is), but that complicated the client code.

**Comment by lewisf** [comment]
@martijnwalraven : I was going to setup a project in swift to make understand this but I figured I could just ask you :) From looking at the generated swift code, it looks like in Swift projects the queries that are generated by apollo-codegen with `__typename` end up being the queries that are executed by the client?

In the flow/typescript stuff right now there isn't a compilation step that produces a new query with `__typename` that clients would execute (and it'd be very difficult to make that the case).
Without explicitly requesting the `__typename` field, client flow/typescript code cannot actually use it within it's runtime for `switch`ing on.

**Comment by lewisf** [comment]
Since javascript/flow targets do always want the `__typename` field in the case of discriminated unions for actual runtime, we probably want the opposite error message there in the case that fragment interfaces are used. If it were implicit, it might be a bit confusing in client code.

For instance given the query:
```graphql
query HeroQuery {
  hero(episode: EMPIRE) {
		name
    friendsConnection {
      friends {
        ...CharacterFragment
      }
    }
  }
}

fragment CharacterFragment on Character {
  name
  __typename

  ... on Human {
    homePlanet
  }
  
  ... on Droid {
    primaryFunction
  }
}
```

In my client code, I'd want to do an explicit switch here

```
  friends.map(friend => {
    switch(friend.__typename) {
      case 'Human': ...
      case 'Droid': ...
    }
  })
```

Without the __typename field being requested in `CharacterFragment` it wouldn't be make sense how it's available to switch on in the code.

**Comment by martijnwalraven** [comment]
@lewisf: Hmmm, you're right, the `apollo-codegen`transform that adds `__typename` doesn't affect the query that `ApolloClient` sends to the server. Hadn't thought of that.

There is an equivalent `addTypename` option on `ApolloClient` however, that I believe is on by default now. Regardless of the default, it can still be turned off, so we may want a way to detect whether it is on when generating type definitions.

It feels a bit messy to me to configure this in multiple places. It seems `persistgraphql` also has its own `--add_typename` option for example.

Maybe this is something we could add to the `.graphqlconfig` file if we add support for it? It allows for client/tool specific configuration options.

**Comment by lewisf** [comment]
@martijnwalraven It does make sense to use whatever configuration is used in general.

For now, maybe a reasonable workaround is probably to just log an error message when generating flow/typescript types if the `__typename` field is needed but isn't in the source query? This would be compatible with any future changes whether or not we decide to add support for `.graphqlconfig`, or add a flag for it.

**Comment by martijnwalraven** [comment]
@lewisf: Hmmm, but I think most people don't currently have `__typename` in their queries, and just use the query transform configured by `addTypename`. So I don't think just looking at the source query will help us here. We could just assume it is on at runtime for now, and always apply the transform in `apollo-codegen` as well maybe. As long as we document that as a requirement I think it is ok for the time being.

**Comment by lewisf** [comment]
@martijnwalraven Okay sounds reasonable to me.

I'm a little concerned about doing a `switch` statement on the `__typename` without explicitly requesting it in the query because it might feel like magic, but that sounds like the right compromise to make here right now and people may just get used to it.

```javascript
class A extends React.Component {
  render() {
    switch (this.props.__typeName) {
      case 'Human':
        return ( ... );
      case 'Droid':
        return ( ... );
    }
  }
}

export default graphql(gql`{
  query Characters {
    characters() {
      ... on Human {
       height
      }

      ... on Droid {
        primaryFunction
      }
    }
  }
`})(A);
```

**Comment by martijnwalraven** [comment]
For the TypeScript code, it seems we could hide the `__typename` magic behind a type guard?

**Comment by clayne11** [comment]
Isn't switching on `__typename` the type guard? How would you hide that and why would you want to?

I would prefer to ask for `__typename` explicitly in these situations. I'm not a fan of magic.

**Comment by martijnwalraven** [comment]
I'm by no means a TypeScript expert, but it seems user-defined type guards are separate functions that narrow the return type when they match? See [here](https://www.typescriptlang.org/docs/handbook/advanced-types.html) (I can't seem to link directly to the right section, so you'll have to scroll down).

**Comment by clayne11** [comment]
Switching on `__typename` is a form of type guard. Type guards have to figure out what type the object is somehow. Using `__typename` is the simplest, cleanest way of doing that IMO.

**Comment by lewisf** [comment]
Following up on this, I think the simplest thing to do is to allow users to specify `__typename` in the query itself. It's recommended here: http://graphql.org/learn/queries/#meta-fields, and even though it's not strictly necessary when a typename query transformer is used, it doesn't feel right to assume a typename query transformer is used by default.

Also, because Flow doesn't have Typescript user-defined type guards, we're going to need the `switch(__typename)` type guard anyway.

We _can_ change this in the future as a toggle-able option, but for now I want to go with the simplest and least surprising solution that's applicable to both Flow and Typescript.

**Comment by clayne11** [comment]
We're going to need the `switch(__typename)` guard in Typescript as well. Regardless of how you define your guards (`switch` vs `function`) you still need to use the `__typename` to discriminate between types.

**Comment by lewisf** [comment]
Yeah that's right. I just landed a change to hopefully get `__typename` added to all the right places. Next step is making sure they get the actual type names as a 'string literal' as `__typename` values

**Comment by clayne11** [comment]
Thanks for all the hard work! I really appreciate it.

**Comment by lewisf** [comment]
This is fixed in #143
